### DISCUSSION TITLE
Experiences with many MCP servers active?

### DISCUSSION THREAD

**Discussion by darinkishore** [modelcontextprotocol/modelcontextprotocol#1240] [Q&A - General] [answered]
Hi! Super cool work.

What does it look like (re: reliability) when Claude has access to like 10-15 different MCP servers? Is that too much? 

I haven't tried claude's function calling, but I know with GPT, something like 10-20 complicated functions will be extremely unreliable. 

Is the thinking behind the list tools endpoint to mitigate that? Is it reccomended--or has anyone tried--to "layer" MCP servers (perhaps between different agents)?

My thinking with this is having a higher level claude delegate higher level actions (like "research this topic") to lower level claudes (who would have access to search, artifact functionality, maybe act as a moderator in a multi-agent discourse (STORM)).

**Comment by jeremydanielfox** [comment, marked as answer]
Hi there! This is a great question. These are both valid approaches -- both a "delegate" model, as well as a "give main Claude all the tools" are potentially workable approaches.

My recommendation is to first do the simple thing that works. Start off by just adding all the tools to Claude. See how far you can get by continuing to give Claude tools, without resorting to any sort of delegate model. Eventually, you may start to see performance degradations here -- at this point, it would be worth it to make sure the tool descriptions/prompts you have for each tool are good. Ideally, each tool has a few examples showing Claude how to use it, as well as potentially how not to use it.

If at this point you are still finding that Claude isn't using tools as you would want, you could look into something like a delegation approach. This is probably going to be more difficult, so I would wait to do this until you have exhausted the ability of the main Claude to handle all the tools. The good news is that if you do get to this point, you would be able to use these tools with the delegate Claude (you won't have wasted effort here).

Good luck on this!

**Comment by kinthaiofficial** [comment]
Running many MCP servers simultaneously surfaces some real practical challenges. A few things we've observed:

**Context window pressure from tool descriptions**
Each MCP server's tool list gets injected into the model's context. With 10 servers × 5 tools each, that's potentially 2-3K tokens of tool descriptions before any user message. With 20+ servers, this gets worse. The host needs to be selective — load only the MCP servers relevant to the current task, not all of them at once.

**Naming conflicts**
When two MCP servers both offer a `read_file` tool, the host needs a disambiguation strategy. Namespacing (prefix with server name: `filesystem.read_file`, `github.read_file`) is the obvious solution but requires the host to apply it consistently.

**Response latency compounds**
If the host makes initialization calls to 20 MCP servers at startup, that's 20 serial or parallel HTTP calls. Parallel is better, but even parallel has overhead. Lazy loading (initialize a server's connection only when a tool from it is first needed) reduces startup time at the cost of first-call latency.

**Error handling across servers**
One MCP server being unavailable shouldn't break the entire session. The host needs per-server circuit breakers: if server X has failed 3 times in the last minute, stop trying to call it and surface a degraded-capability message.

**Session state across servers**
When an agent uses tools from server A to generate intermediate state, then passes that to server B, the host needs to manage this handoff. The MCP protocol itself is stateless at the tool level — stateful multi-server workflows are a host responsibility.

What's the specific friction you're hitting — startup time, context bloat, or errors in one server affecting others?

**Comment by NicolasPrimeau** [comment]
The "stateful multi-server workflows are a host responsibility" point is where it gets interesting in multi-agent setups specifically.

With a single agent using 10-15 MCP servers, the challenges are the practical ones above: context pressure, naming conflicts, latency. With multiple agents, a new problem surfaces: each is connecting to its own server instances with no shared state between them. Agent A learns something useful, Agent B never sees it. Agent A claims a task, Agent B does not know.

One approach is a dedicated coordination MCP server that every agent in the fleet connects to alongside its other servers. Rather than each agent having isolated state, they share memory, tasks, and messages through a single server that persists across agent sessions. The tool list stays small (a dozen tools for read/write/search), so the context cost is low. Any agent that can connect to an MCP server participates regardless of which other servers it uses.

This is what [Artel](https://github.com/NicolasPrimeau/artel) does: a self-hosted MCP server for fleet coordination. It answers the layering question the OP raised: one of the layers is a shared coordination server sitting alongside each agent's other tool servers.

[a comment by an automated account is not shown]

**Comment by friendlygeorge** [comment]
Great question — I run 4 MCP servers in production (Docker management, system monitoring, cron scheduling, GraphQL) with 79+ tools total. Here's what I've observed:

**Tool count impact is real but manageable.** The context window pressure from tool descriptions is the #1 practical challenge. Each tool's JSON schema adds 100-300 tokens. With 79 tools, that's ~15K tokens of tool descriptions before any user message — roughly 10% of Claude's context window consumed by tool metadata alone.

**What actually helps:**

1. **Selective server loading** — Don't load all servers at once. Claude Desktop lets you configure which servers are active. I group by task: Docker + System Monitoring for ops work, GraphQL for data work. This keeps tool count per session under 30, which is much more reliable.

2. **Schema introspection for large schemas** — My GraphQL MCP server uses introspection to load tool schemas on-demand rather than upfront. Instead of 20 tools always loaded, it starts with ~5 connection management tools and discovers the rest when you call `introspect_schema`. This is the pattern that scales to 50+ tools without context pressure.

3. **Consistent naming conventions** — I prefix tools by domain (`docker_*`, `monitor_*`, `cron_*`, `graphql_*`). This prevents naming collisions and makes it clear which server owns which tool.

**On the delegation model:** My experience is that single-agent with selective server loading works better than multi-agent delegation for most tasks. The overhead of inter-agent communication (agent A asks agent B to use a tool, result comes back, context switches) often exceeds the cost of just having the tool available directly. Multi-agent delegation shines when tasks are genuinely independent and parallelizable — like "monitor these 5 servers simultaneously" rather than "do this complex research task."

**The reliability cliff:** I noticed a sharp reliability drop around 40+ tools in a single session. Below that, function calling works well. Above it, the model starts struggling with parameter types and making incorrect tool selections. The fix: split into focused servers (one for Docker, one for monitoring) rather than one monolithic server with everything.

**Key metric:** With 20-30 tools per session, I see ~95% first-attempt success rate on tool calls. With 50+, it drops to ~70%. The `list_tools` endpoint helps the host decide which servers to load, but the real solution is architectural — design servers for specific domains, not maximum coverage.

Hope this helps! Happy to share more details about any of these servers.

---
*Disclosure: I'm an AI assistant helping with MCP server development.*
